chore: enable strict mode for workspace-minimap

Namatuzio opened the pull request:
## The basics

- [x] I [validated my changes](https://developers.google.com/blockly/guides/contribute/samples#making_and_verifying_a_change)

## The details
### Resolves

Fixes #2031 

### Proposed Changes

* Updates the workspace-minimap plugin's type checks in accordance with TypeScript's [strict-mode](https://www.typescriptlang.org/tsconfig#strict)
* Updated the following files:
  * `plugins\workspace-minimap\src\minimap.ts`
  * `plugins\workspace-minimap\src\focus_region.ts`
  * `plugins\workspace-minimap\test\index.ts`
  * `plugins\workspace-minimap\test\minimap_tests.mocha.js`

### Reason for Changes

Strict mode is a great way to ensure strong code consistency and maintain high code standards. It allows for fewer faults within code to be possible by patching potential oversights.

### Test Coverage

`npm run test` results:
![image](https://github.com/google/blockly-samples/assets/55001591/6122db22-945c-410b-bd6f-79f6f87d678a)


### Documentation

### Additional Information

There are 2 warnings when running `npm run test` from a change I added. For some reason when I ran the tests, the tests wouldn't have access to the document object part of the `DOM`, so I added the following to the top of the tester file which seemed to have fixed it:

```ts
const jsdom = require('jsdom');
const { JSDOM } = jsdom;

const { document } = (new JSDOM('')).window;
global.document = document;
```

Namatuzio committed:
Fixed workscpace-minimap/focus_region.ts type errors

Added various type checks and changes to focus_region.ts which follows proper standards for the TypeScript "strict-mode" compiler checks.

Namatuzio committed:
fixed type errors in workspace-minimap/minimap.ts

Fixed all type error in ordinance with TypeScript's "strict-mode"

Namatuzio committed:
Updated workspace-minimap to utilize TS strict-mode

Fully updated and fixed all type errors after enabling strict-mode.

Namatuzio committed:
Updated index.ts to follow strict-mode rules

Namatuzio committed:
fix lint errors in index.ts

Namatuzio committed:
Merge branch 'google:master' into enable-strict-workspace-minimap

Namatuzio committed:
Updated testing file to coordinate with strict-mode

Namatuzio committed:
Merge branch 'enable-strict-workspace-minimap' of https://github.com/Namatuzio/blockly-samples into enable-strict-workspace-minimap

BeksOmega commented on plugins/workspace-minimap/src/focus_region.ts:
```diff
@@ -27,10 +27,10 @@ const borderRadius = 6;
-  private svgGroup: SVGElement;
-  private rect: SVGElement;
-  private background: SVGElement;
+  private onChangeWrapper: ((e: Blockly.Events.Abstract) => void) | null;
```
We can assign this in the constructor, so please remove the `| null` here.

BeksOmega commented on plugins/workspace-minimap/src/focus_region.ts:
```diff
@@ -45,6 +45,10 @@ export class FocusRegion {
     private minimapWorkspace: Blockly.WorkspaceSvg,
   ) {
     this.id = String(Math.random()).substring(2);
+    this.onChangeWrapper = (e: Blockly.Events.Abstract) => {}; 
```
Instead of assigning an empty method, can you assign `this.onChange.bind()` here, and remove the later assignment?

BeksOmega commented on plugins/workspace-minimap/src/focus_region.ts:
```diff
@@ -45,6 +45,10 @@ export class FocusRegion {
+    this.onChangeWrapper = (e: Blockly.Events.Abstract) => {}; 
+    this.svgGroup = document.createElementNS('http://www.w3.org/2000/svg', 'g');
+    this.rect = document.createElementNS('http://www.w3.org/2000/svg', 'rect');
+    this.background = document.createElementNS('http://www.w3.org/2000/svg', 'rect'); 
```
If these are `| null` we don't need to assign them in the constructor. Can you remove these?

BeksOmega commented on plugins/workspace-minimap/src/focus_region.ts:
```diff
@@ -187,9 +191,9 @@ export class FocusRegion {
-    this.rect.setAttribute('height', height.toString());
+    this.rect?.setAttribute('transform', `translate(${left},${top})`);
+    this.rect?.setAttribute('width', width.toString());
+    this.rect?.setAttribute('height', height.toString());
```
Instead of using `?` can you throw an error at the top of this function if `this.rect` is not initialized?

Like:
```
if (!this.rect) {
  throw new Error('The focus region must be initialized (`init`) before calling `update`');
}
```

BeksOmega commented on plugins/workspace-minimap/src/minimap.ts:
```diff
@@ -44,6 +44,9 @@ export class Minimap {
     this.primaryWorkspace = workspace;
+    this.minimapWorkspace = new Blockly.WorkspaceSvg(new Blockly.Options({}));
+    this.focusRegion = new FocusRegion(this.primaryWorkspace, this.minimapWorkspace);
+    this.minimapWrapper = null;
```
No need to assign in the constructor if it is `null` by default. Please remove this.

BeksOmega commented on plugins/workspace-minimap/src/minimap.ts:
```diff
@@ -97,12 +100,16 @@ export class Minimap {
-      this.onClickUp,
-    );
+    if(primaryInjectParentDiv === null) {
+      throw new Error("primaryInjectParentDiv is null");
```
```suggestion
    if (!primaryInjectionDiv) {
      throw new Error('The workspace must be injected into the page before the minimap can be initalized');
```
Can you also move this to the top of the `init` method? The `this.onMouseUpWrapper` assignment can be extracted from the `else` and left where it is.

BeksOmega commented on plugins/workspace-minimap/src/positioned_minimap.ts:
```diff
@@ -179,19 +179,23 @@ export class PositionedMinimap
-    Blockly.svgResize(this.minimapWorkspace);
+    if(injectDiv.parentElement === null) {
+      return;
+    } else {
```
No need to wrap in an `else` if you are early returning. Please remove the `else` part of this clause.

BeksOmega commented on plugins/workspace-minimap/test/minimap_tests.mocha.js:
```diff
@@ -9,6 +9,11 @@ const assert = chai.assert;
+const { JSDOM } = jsdom;
+
+const { document } = (new JSDOM('')).window;
+global.document = document;
```
I'll look into this and see if it's necessary!

BeksOmega reviewed (changes requested):
Thank you for your work on this! This looks like a great start :D

Namatuzio committed:
Merge branch 'google:master' into enable-strict-workspace-minimap

Namatuzio committed:
Added requested changes

Added the changes request under the PR

BeksOmega commented on plugins/workspace-minimap/src/focus_region.ts:
```diff
@@ -206,4 +211,4 @@ Blockly.Css.register(`
   fill: #e6e6e6;
 }
-`);
+`);
```
Can you add a newline at the end of all the files that have the ⊖ symbol? (here and below)

BeksOmega commented on plugins/workspace-minimap/src/positioned_minimap.ts:
```diff
@@ -179,19 +179,22 @@ export class PositionedMinimap
-    style.top = `${this.top}px`;
-    style.left = `${this.left}px`;
-    Blockly.svgResize(this.minimapWorkspace);
+    if(injectDiv.parentElement === null) {
```
Can you fix the formatting here?

BeksOmega commented on plugins/workspace-minimap/src/minimap.ts:
```diff
@@ -44,6 +44,8 @@ export class Minimap {
   constructor(workspace: Blockly.WorkspaceSvg) {
     this.primaryWorkspace = workspace;
+    this.minimapWorkspace = new Blockly.WorkspaceSvg(new Blockly.Options({}));
+    this.focusRegion = new FocusRegion(this.primaryWorkspace, this.minimapWorkspace);
```
I think we actually want to make these nullable. Constructing them with dummy values and then reconstructing them with real values is computationally expensive.

Can you remove this and then update the properties to be nullable throughout?

BeksOmega commented on plugins/workspace-minimap/src/minimap.ts:
```diff
@@ -58,58 +60,62 @@ export class Minimap {
-      renderer: this.primaryWorkspace.options.renderer,
-    });
+    if (!primaryInjectParentDiv) {
+      throw new Error('The workspace must be injected into the page before the minimap can be initalized');
```
Can you move this to the top of the `init` method? You can also remove the `else` clause, because throwing an error early-returns.

Namatuzio committed:
Added feedback changes

Namatuzio committed:
Removed JSDOM test code

Removed the JSDOM code found in `plugins/workspace-minimap/test/minimap-test.mocha.js` as it was unnecessary due to updates to the code.

Namatuzio committed:
format files using prettier
